backends/avx/ceed-avx-basis.c

```diff
@@ -0,0 +1,362 @@
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <string.h>
+#include <immintrin.h>
+#include "ceed-avx.h"
+
+// Contracts on the middle index
+// NOTRANSPOSE: V_ajc = T_jb U_abc
+// TRANSPOSE:   V_ajc = T_bj U_abc
+// If Add != 0, "=" is replaced by "+="
+static int CeedTensorContract1_Avx(Ceed ceed, CeedInt A, CeedInt B,
+                                   CeedInt C, CeedInt J,
+                                   const CeedScalar *restrict t,
+                                   CeedTransposeMode tmode, const CeedInt Add,
+                                   const CeedScalar *restrict u, CeedScalar *restrict v) {
+  CeedInt tstride0 = B, tstride1 = 1;
+  if (tmode == CEED_TRANSPOSE) {
+    tstride0 = 1; tstride1 = J;
+  }
+
+  if (!Add)
+    for (CeedInt q=0; q<A*J*C; q++)
+      v[q] = (CeedScalar) 0.0;
+
+  for (CeedInt a=0; a<A; a++)
+    for (CeedInt b=0; b<B; b++)
+      for (CeedInt j=0; j<J; j++) {
+        CeedScalar tq = t[j*tstride0 + b*tstride1];
+        for (CeedInt c=0; c<C; c++)
+          v[(a*J+j)*C+c] += tq * u[(a*B+b)*C+c];
+      }
+  return 0;
+}
+
+static int CeedTensorContract8_Avx(Ceed ceed, CeedInt A, CeedInt B,
+                                   CeedInt C, CeedInt J,
+                                   const CeedScalar *restrict t,
+                                   CeedTransposeMode tmode, const CeedInt Add,
+                                   const CeedScalar *restrict u, CeedScalar *restrict v) {
+  CeedInt tstride0 = B, tstride1 = 1;
+  if (tmode == CEED_TRANSPOSE) {
+    tstride0 = 1; tstride1 = J;
+  }
+
+  const int JJ = 4, CC=8;
+  if (C % CC) return CeedError(ceed, 2, "Tensor [%d, %d, %d]: last dimension not divisible by %d", A, B, C, CC);
+
+  if (!Add) {
+    for (CeedInt q=0; q<A*J*C; q++) {
+      v[q] = (CeedScalar) 0.0;
+    }
+  }
+
+  for (CeedInt a=0; a<A; a++) {
+    // Blocks of 4
+    for (CeedInt j=0; j<(J/JJ)*JJ; j+=JJ) {
+      for (CeedInt c=0; c<C; c+=CC) {
+        __m256d vv[JJ][CC/4]; // Output tile to be held in registers
+        for (CeedInt jj=0; jj<JJ; jj++)
+          for (CeedInt cc=0; cc<CC/4; cc++)
+            vv[jj][cc] = _mm256_loadu_pd(&v[(a*J+j+jj)*C+c+cc*4]);
+
+        for (CeedInt b=0; b<B; b++) {
+          for (CeedInt jj=0; jj<JJ; jj++) { // unroll
+            __m256d tqv = _mm256_set1_pd(t[(j+jj)*tstride0 + b*tstride1]);
+            for (CeedInt cc=0; cc<CC/4; cc++) { // unroll
+              vv[jj][cc] += _mm256_mul_pd(tqv,
+                              _mm256_loadu_pd(&u[(a*B+b)*C+c+cc*4]));
```

Comment:
I haven't checked if * works - I just saw this as the typical workaround with the error @valeriabarra caught.

I can look into different values of CC and JJ. Would this be seen in perf?

Reply:
Yeah, I like to use perf because it's pretty precise for that sort of thing and it's easy to map from source code to assembly. IACA might also be useful. https://stackoverflow.com/questions/26021337/what-is-iaca-and-how-do-i-use-it

Reply:
Modulo the portability issue that will need to be fixed, it's up to you whether to merge this now and work on further optimizations in a separate PR or to try to finish the job in this PR.